File: sockpuppet.py
#!/usr/bin/python

import json
import random
import string
import logging

import tornado.httpclient
import tornado.gen
import tornado.options
import tornado.web
import tornado.websocket


def parse_payload(data):
    try:
        body = None
        frame = json.loads(data)
        if isinstance(frame, list) and len(frame) > 0:
            frame = frame[0]
            body = json.loads(frame.get('body', "{}"))

        return frame, body

    except Exception as ex:
        logging.exception(ex)

    return None, None


@tornado.gen.coroutine
def connect_and_read_websocket():

    url = tornado.options.options.ws_addr
    try:
        logging.info("connecting to: %s", url)
        w = yield tornado.websocket.websocket_connect(url, connect_timeout=5)
        logging.info("connected to %s, waiting for messages", url)
    except Exception as ex:
        logging.error("couldn't connect, err: %s", ex)
        return

    while True:
        payload = yield w.read_message()
        if payload is None:
            logging.error("uh oh, we got disconnected")
            return

        if payload[0] == "o":
            cookie = ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(32))
            msg = json.dumps(['{"action":"login", "client_app":"hermes.push", "cookies":{"nyt-s":"%s"}}' % cookie])
            w.write_message(msg.encode('utf8'))
            logging.info("sent cookie: %s", cookie)

        elif payload[0] == 'h':
            logging.info('ping')

        elif payload[0] == 'a':
            frame, body = parse_payload(payload[1:])
            if not frame:
                logging.warn("unknown paylod: %s", payload.decode('utf8'))
                continue

            if frame['product'] == "core" and frame['project'] == "standard":
                logging.info("Logged in ok")

            elif frame['product'] == "hermes" and frame['project'] == "push" and body:
                logging.warn("NEW! %s: %s", body['label'], body['title'])

        else:
            logging.warn("unknown paylod: %s", payload.decode('utf8'))


if __name__ == '__main__':
    tornado.options.define(name="ws_addr", type=str, help="Address of the websocket host to connect to.")
    tornado.options.parse_command_line()
    if not tornado.options.options.ws_addr:
        tornado.options.print_help()
        exit()

    tornado.ioloop.IOLoop.instance().run_sync(connect_and_read_websocket)

File: sockpuppet.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"strings"
	"time"

	"code.google.com/p/go.net/websocket"
)

var (
	wsAddr = flag.String("ws_addr", "", "Address of the host to connect to")
)

type ServerMessage []struct {
	UUID        string `json:"uuid"`
	Timestamp   string `json:"timestamp"`
	Region      string `json:"region"`
	Zone        string `json:"zone"`
	Product     string `json:"product"`
	Project     string `json:"project"`
	Environment string `json:"environment"`
	Type        string `json:"type"`
	Body        string `json:"body,omitempty"`
}

type MessageBody struct {
	ID           int    `json:"id"`
	Title        string `json:"title"`
	Status       string `json:"status"`
	Version      int    `json:"version"`
	SubType      string `json:"sub_type"`
	Label        string `json:"label"`
	StartTime    int    `json:"start_time"`
	EndTime      int    `json:"end_time"`
	LastModified int    `json:"last_modified"`

	Links []struct {
		URL         string `json:"url"`
		Count       int    `json:"count"`
		ContentID   string `json:"content_id"`
		ContentType string `json:"content_type"`
		Offset      int    `json:"offset"`
	} `json:"links"`
}

func main() {
	rand.Seed(time.Now().Unix())
	flag.Parse()

	if *wsAddr == "" {
		fmt.Println("Need to provide a valid host via --ws_addr=\"\"")
		return
	}

	addr := *wsAddr
	switch {
	case strings.HasPrefix(addr, "ws://"):
		addr = strings.Replace(*wsAddr, ".com./", ".com.:80/", 1)
	case strings.HasPrefix(addr, "wss://"):
		addr = strings.Replace(*wsAddr, ".com./", ".com.:443/", 1)
	}

	ws, err := websocket.Dial(addr, "", "http://www.nytimes.com/")
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("Connected to %s", addr)

	var msgBuf = make([]byte, 4096)
	for {
		bufLen, err := ws.Read(msgBuf)
		if err != nil {
			log.Printf("read err: %s", err)
			break
		}

		if bufLen < 1 {
			continue
		}

		switch msgBuf[0] {
		case 'o':
			// reply to the login request
			cookie := randCookie()
			msg := fmt.Sprintf(`["{\"action\":\"login\", \"client_app\":\"hermes.push\", \"cookies\":{\"nyt-s\":\"%s\"}}"]`, cookie)
			_, err := ws.Write([]byte(msg))
			if err != nil {
				log.Fatal(err)
			}
			log.Printf("Sent cookie: %s\n", cookie)

		case 'h':
			// keep-alive
			log.Println("ping")

		case 'a':
			// some JSON encoded data, let's decode it!
			msg, err := decodeServerMessage(msgBuf[1:bufLen])
			if err != nil {
				log.Printf("no good: %s", err)
				continue
			}

			// response to the login message?
			if msg[0].Product == "core" && msg[0].Project == "standard" {
				log.Printf("Logged in ok")
				continue
			}

			// possibly breaking news?
			if msg[0].Product == "hermes" && msg[0].Project == "push" && len(msg[0].Body) > 0 {
				body, err := decodeMessageBody(msg[0].Body)
				if err != nil {
					log.Printf("decoding err: %s    msg: %s", err, msg)
					continue
				}
				log.Printf("NEW! %s  %s: %s", body.SubType, body.Label, body.Title)
			}

		default:
			log.Printf("No idea what this is: %s\n", msgBuf)
		}
	}

	log.Printf("Exiting")
}
func decodeMessageBody(body string) (res MessageBody, err error) {
	err = json.Unmarshal([]byte(body), &res)
	return
}

func decodeServerMessage(buf []byte) (res ServerMessage, err error) {
	err = json.Unmarshal(buf, &res)
	if err != nil {
		return
	}
	if len(res) == 0 {
		err = fmt.Errorf("shouldn't be empty")
	}
	return
}

func randCookie() string {
	letters := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890")
	b := make([]rune, 30)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}
